### PR TITLE
[NEW] Global message search (beta: disabled by default)

### LINKED ISSUE #1615: Global search across channels
As far as I can tell, search right now functions on a per-channel basis. Is that true? Is there a way to search all chat history globally? This seems extremely useful, and probably not hard to implement.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28998888-global-search-across-channels?utm_campaign=plugin&utm_content=tracker%2F16890630&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F16890630&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by rodrigok**
# Description

It adds a new setting to enable the global search beta feature allowing users to search messages using the contextual bar (right bar) and choose to receive results from the current room only or from all rooms that he already joined.

## Settings
- Message_GlobalSearch: Boolean (default false)

----

@RocketChat/core 

Based on https://github.com/RocketChat/Rocket.Chat/pull/7628 and https://github.com/RocketChat/Rocket.Chat/pull/7274 thanks @cyberhck and @savikko

Closes #7628 
Closes #7274 
Closes #1615

screen shot 2018-02-12 at 11 52 56 am

image

@karlprieb we need to improve the checkbox ui at search contextual bar

### CODE REVIEW

**Comment by meghalagrawal** [comment]
Can I work on this issue?

**Comment by geekgonecrazy** [comment]
What about combining this with https://github.com/RocketChat/Rocket.Chat/pull/8714 ?

Seems like proper implementation here would be to plug this up as providers.

Default provider of search being this one.  But being able to plug in external provider would be the perfect solution.

It seems like we really should do this properly instead of just merging alone?  Or if we merge alone we need to make sure and let everyone know what are plans are?

@RocketChat/core thoughts?

> Can I work on this issue?
@meghalagrawal this is a PR not an issue :)

**Comment by cyberhck** [comment]
wow, the UI looks million times better than what I saw last time. (again, I'm poor in design)

**Comment by rodrigok** [comment]
@geekgonecrazy would be awesome for a future implementation, the idea here is to have a working solution to solve some team issues while trying to find the correct message, wasn't developed to be the best option. But I'd lide to see a better implementation in the future to improve this.

**Comment by cyberhck** [comment]
In the future, yes, the provider option should work, maybe one wants different types of search, maybe someone might want to index stuff on elastic search, and maybe keywords and stuff, provider should be quite good. But for now since this is a highly asked feature, I think it's a good idea to get this first, but it's upto maintainers, so.

**Comment by geekgonecrazy** [comment]
Will be having conversations to attempt the provider approach after this PR.  👍 might as well include this behind beta flag and get some bugs found, reported and fixed.

**Comment by tkurz** [comment]
Hi all! As you might already heard from @mrsimpson we are currently working on a search provider concept for Rocketchat. As we are working with solr for a long time, we already implemented a search provider for solr (which we now slightly adapt to fit the provider concept). You can see the current status in a nice gif (thx @mrsimpson) here https://github.com/redlink-gmbh/chatpal-search/pull/14#issuecomment-364222320. We will have a alpha version of the concept mid of next week and would looking forward to present and discuss it with you!

File: packages/rocketchat-ui-flextab/client/tabs/messageSearch.js [hunk 1]
```diff
@@ -10,11 +10,22 @@ Meteor.startup(function() {
...
+			FlowRouter.goToRoomById(message.rid);
+			// RocketChat.MessageAction.hideDropDown();
+
 			if (window.matchMedia('(max-width: 500px)').matches) {
 				Template.instance().tabBar.close();
 			}
 
-			RoomHistoryManager.getSurroundingMessages(message, 50);
+			window.setTimeout(() => {
+				RoomHistoryManager.getSurroundingMessages(message, 50);
```
Comment (cyberhck) [inline packages/rocketchat-ui-flextab/client/tabs/messageSearch.js]: I see this is still there, I was trying so hard to find if there's any other way than setting a timeout, I can't remember if FlowRouter.goToRoomById returns any promise/callback

**Comment by tkurz** [comment]
Hi guys! 
As you might have seen I create a pull request #10110 which integrates a provider concept liked discussed in  #8714 with your search. I would be great if you could have a closer look at it!
I hope you like the approach!

**Comment by paulharris** [comment]
Surely this feature should be finished properly and enabled by default by now?

**Comment by eishta-christy** [comment]
Hi, I'm new here, and very much interested on the developing of the global search in Rocket.Chat. My team is using rocket chat and for the short period of time it was available, it was really helpful! @tkurz could you help me to understand the current state?

**Pull request merged**